### PR TITLE
Optimize segmentwise recompression

### PR DESCRIPTION
**Pull request by antekresic**
With the recent change to compression schema, we can optimize segmentwise recompression and target only batches which are affected to be recompressed or compress tuples which don't belong to any batch without any decompression. This means narrowing down the amount of that we need to decompress and recompress, ultimately speeding up the operation and generating less bloat in the compressed chunk.

Benchmark with improvements (2-3x in certain cases):
https://grafana.ops.savannah-dev.timescale.com/d/fasYic_4z/compare-akuzm?orgId=1&var-branch=All&var-run1=3959&var-run2=3960&var-threshold=0&var-use_historical_thresholds=true&var-threshold_expression=2.5%20%2A%20percentile_cont%280.90%29&var-exact_suite_version=false

### CODE REVIEW

**Comment by erimatnor** [comment]
Do we have benchmarks to confirm that this is faster?

File: tsl/src/compression/recompress.c [hunk 1]
```diff
@@ -0,0 +1,725 @@
...
+
+	ts_feature_flag_check(FEATURE_HYPERTABLE_COMPRESSION);
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+	Chunk *chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (!ts_chunk_is_partial(chunk))
+	{
+		int elevel = if_not_compressed ? NOTICE : ERROR;
+		elog(elevel,
+			 "nothing to recompress in chunk %s.%s",
```
Comment (akuzm) [inline tsl/src/compression/recompress.c:67]: This is going to have sqlstate XX000 that's reserved for program logic errors, let's use some other errcode.

File: tsl/src/compression/recompress.c [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+	Oid uncompressed_chunk_id = uncompressed_chunk->table_id;
+
+	/*
+	 * only proceed if status in (3, 9, 11)
+	 * 1: compressed
+	 * 2: compressed_unordered
+	 * 4: frozen
+	 * 8: compressed_partial
+	 */
+	if (!ts_chunk_is_compressed(uncompressed_chunk) && ts_chunk_is_partial(uncompressed_chunk))
```
Comment (akuzm) [inline tsl/src/compression/recompress.c:99]: This condition reads a little weird, how can chunk be uncompressed and partially compressed at the same time? Might be not a topic for this PR, but we might have to rename `ts_chunk_is_compressed` to something like `ts_chunk_is_fully_compressed`.

File: tsl/src/compression/recompress.c [hunk 3]
```diff
@@ -0,0 +1,725 @@
...
+		elog(ERROR,
```
Comment (akuzm) [inline tsl/src/compression/recompress.c:100]: Looks like this still can be triggered by a user calling it for an unsuitable chunk? Needs better errcode then.

File: tsl/src/compression/recompress.c [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+				(errmsg("cleared chunk status for recompression: \"%s.%s\"",
+						NameStr(uncompressed_chunk->fd.schema_name),
+						NameStr(uncompressed_chunk->fd.table_name))));
+
+	ereport(DEBUG1,
+			(errmsg("acquiring locks for recompression: \"%s.%s\"",
+					NameStr(uncompressed_chunk->fd.schema_name),
+					NameStr(uncompressed_chunk->fd.table_name))));
+	/* lock both chunks, compressed and uncompressed */
+	/* TODO: Take RowExclusive locks instead of AccessExclusive */
```
Comment (akuzm) [inline tsl/src/compression/recompress.c:127]: Still todo? Might be good to add more explanation to the comment on what else is needed to move to RowExclusive.

File: tsl/src/compression/recompress.c [hunk 5]
```diff
@@ -0,0 +1,725 @@
...
+	Relation uncompressed_chunk_rel = table_open(uncompressed_chunk->table_id, ExclusiveLock);
+	Relation compressed_chunk_rel = table_open(compressed_chunk->table_id, ExclusiveLock);
+
+	/*
+	 * Calculate and add the column dimension ranges for the src chunk. This has to
```
Comment (akuzm) [inline tsl/src/compression/recompress.c:132]: Might be good to mention that this is for the `enable_chunk_skipping()` functionality, the name "column stats" is very generic so I was unsure.

File: tsl/src/compression/recompress.c [hunk 6]
```diff
@@ -0,0 +1,725 @@
...
+
+	Tuplesortstate *batch_tuplesortstate = tuplesort_begin_heap(uncompressed_rel_tupdesc,
+																n_keys,
+																sort_keys,
+																sort_operators,
+																sort_collations,
+																nulls_first,
+																maintenance_work_mem,
+																NULL,
+																false);
```
Comment (akuzm) [inline tsl/src/compression/recompress.c:237]: This is always inside one segment group, should it sort only by orderby columns?

The name is not very precise, maybe something like `recompressing_tuplesort`? In the sense that these are the tuples of the batches that we're currently recompressing.

File: tsl/src/compression/recompress.c [hunk 7]
```diff
@@ -0,0 +1,725 @@
...
+					NameStr(uncompressed_chunk->fd.schema_name),
+					NameStr(uncompressed_chunk->fd.table_name))));
+
+	/* Setting up scankeys */
+	ScanKeyData *index_scankeys = palloc(sizeof(ScanKeyData) * num_segmentby);
+	ScanKeyData *orderby_scankeys = palloc(sizeof(ScanKeyData) * num_orderby * 2);
+	create_segmentby_scankeys(settings, index_rel, compressed_chunk_rel, index_scankeys);
+	create_orderby_scankeys(settings, index_rel, compressed_chunk_rel, orderby_scankeys);
+
+	Tuplesortstate *segment_tuplesortstate = tuplesort_begin_heap(uncompressed_rel_tupdesc,
```
Comment (akuzm) [inline tsl/src/compression/recompress.c:219]: This is just the entire uncompressed part of the chunk sorted in compression order, right? Maybe something like `new_tuples` or "incoming"... Also could call it "uncompressed" but this word might be too heavily loaded. `uncompressed_part_tuplesort`?

File: tsl/src/compression/recompress.c (hunk 2 above)
Comment (antekresic) [reply tsl/src/compression/recompress.c:99]: Yeah, that check seems to check for two bitfields which should always accompany each other (compressed and partially compressed). Having partially compressed without the compressed bit should be an internal error.

Maybe `Ensure` is more suitable for this check?

File: tsl/src/compression/recompress.c (hunk 3 above)
Comment (antekresic) [reply tsl/src/compression/recompress.c:100]: No, this shouldn't not happen in regular execution. That's why I think `Ensure` would be better suited here.

File: tsl/src/compression/recompress.c (hunk 4 above)
Comment (antekresic) [reply tsl/src/compression/recompress.c:127]: Adding bit more context to the TODO.

File: tsl/src/compression/recompress.c (hunk 6 above)
Comment (antekresic) [reply tsl/src/compression/recompress.c]: Renamed to `recompress_tuplesortstate` and added a comment for clearing up the usage pattern.

File: tsl/src/compression/recompress.c (hunk 7 above)
Comment (antekresic) [reply tsl/src/compression/recompress.c]: Renamed to `input_tuplesortstate` and added comment to clear up its intended use.

File: tsl/src/compression/recompress.c (hunk 1 above)
Comment (antekresic) [reply tsl/src/compression/recompress.c]: Added a specific error code which better reflects the nature of the error.

File: tsl/src/compression/recompress.c [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+ * compressed chunk will not be recreated but modified in place.
+ *
+ * 0 uncompressed_chunk_id REGCLASS
+ * 1 if_not_compressed BOOL = false
+ */
+Datum
+tsl_recompress_chunk_segmentwise(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_not_compressed = PG_ARGISNULL(1) ? true : PG_GETARG_BOOL(1);
```
Comment (svenklemm) [inline tsl/src/compression/recompress.c:57]: The default is true here while in the function comment it is stated as false

**Review by svenklemm — APPROVED** [review]

File: tsl/src/compression/recompress.c (hunk 8 above)
Comment (antekresic) [reply tsl/src/compression/recompress.c:57]: Will address this in a follow-up PR since its current state and has very little to do with the change here.

**Pull request merged**